piptools/repositories/pypi.py

```diff
@@ -99,8 +112,9 @@ def get_dependencies(self, ireq):
         dependencies (also InstallRequirements, but not necessarily pinned).
         They indicate the secondary dependencies for the given requirement.
         """
-        if not (ireq.editable or is_pinned_requirement(ireq)):
-            raise TypeError('Expected pinned or editable InstallRequirement, got {}'.format(ireq))
+
+        if ireq in self._get_dependencies_cache:
+            return self._get_dependencies_cache[ireq]
```

Comment:
Is this a perf optimization? I really like to avoid adding perf logic in the middle of the processing, as it complicates the logic and makes it hard to follow along what's happening. Let's solve it first, and optimize for performance later. If this is not strictly necessary for the functionality, please remove it from this PR.